### DISCUSSION TITLE
WASM Multithreading

### DISCUSSION THREAD

**Discussion by BenMcLean** [unoplatform/uno#19035] [Q&A]
For several years, I've been writing a .NET Standard 2.0 library (using PolySharp for newer C# language features) that does some graphics processing on the CPU, which means many operations it does are CPU-bound. When it came time to add multi-threading, I went through the codebase and just replaced a lot of `Select` queries with this:
```csharp
/// <summary>
/// Parallelizes the execution of a Select query while preserving the order of the source sequence.
/// </summary>
public static List<TResult> Parallelize<TSource, TResult>(this IEnumerable<TSource> source, Func<TSource, TResult> selector) => [.. source
	.Select((element, index) => (element, index))
	.AsParallel()
	.Select(sourceTuple => (result: selector(sourceTuple.element), sourceTuple.index))
	.OrderBy(resultTuple => resultTuple.index)
	.AsEnumerable()
	.Select(resultTuple => resultTuple.result)];
```
Just replacing `Select` with `Parallelize` made the same operations take a small fraction of the time they used to. It works great whenever you have a series of CPU-bound tasks that you know in advance don't depend on each other or on anything that changes to get done.

Now I'd like to give my library a web front-end for people to use and host the web front-end for free on Github Pages. So naturally, the first tool I reached for was Blazor WASM because that would be the Microsoft way to do this. The Parallelize method I wrote works on Blazor WASM, except single-threaded because [Blazor WASM only provides a single-threaded runtime](https://github.com/dotnet/aspnetcore/issues/17730). Being single-threaded means it blocks the UI which is bad ... but the user's patience is eventually rewarded with the result they requested so that's good.

But then I read that [Uno Platform has multi-threading support on WASM](https://platform.uno/docs/articles/external/uno.wasm.bootstrap/doc/features-threading.html) but I'm not sure I fully understand the requirements. Would a PLINQ method like this be able to actually use multi-threading? Apparently it requires a special header to be added to the site: so is this something I could still host for free on Github Pages? When making multi-threaded code, what stuff is allowed vs not allowed on WASM?

The real point / heart of my question is that I am considering whether I should port my front-end from Blazor WASM to Uno or just start over coding the front-end in Uno solely in order to get the multi-threading feature. Is that worth it? I'd hate to invest many hours in learning the Uno platform just for this only to realize that multi-threading for my application would be untenable.

**Comment by jeromelaban** [comment]
Thanks for the question.

Wasm threading is experimental overall, is a best effort support in .NET 8, and [is not supported in .NET 9 on Uno's side](https://platform.uno/docs/articles/external/uno.wasm.bootstrap/doc/using-the-bootstrapper.html#threading) because of structural changes made by the runtime. We expect this to change in .NET 10, however.

**Reply by BenMcLean** [reply]
So if I were to do this, I'd need to make my Uno project in .NET 8. OK, that's good to know, but which multi-threading techniques in the code would (at least most likely) work and which wouldn't? Is this multi-threading support going through some kind of special JavaScript interop class to make the threads or is it just implementing it in the runtime to work with normal C# async/await methods? Would PLINQ like I showed work?

**Reply by jeromelaban** [reply]
Threading that does not use any javascript interop is generally accepted in net8, you'd otherwise need to go back to the UI thread first. You can use any async method that does actual parallel CPU-bound work.

The best is to try it out and see what works for you.

**Comment by tobizk04** [comment]
Hello Jerome (@jeromelaban).
What are the problems to support multithreading on uno wasm?
At the moment my team implement an 2D app for ECAD data.
Multithreading is essential for us. We render 2D with skiasharp.

It works well with .net8 but if the future is unsafe for uno and multithreading it is a big risk for us.
So we must decide if we go forward with uno platform.
What do you mean?
Could you help us?
Could we help you to get multithreading support on uno faster?

**Reply by francoistanguay** [reply]
Hi @tobizk04 , there's definitely a future for Wasm/Multi-Threading support with Uno. net9 brought a minor setback but we're exploring various strategies.

I'd love to explore some of these options with you/your team and better understand your project needs.

Feel free to reach out by email francois - at - platform dot uno so we can schedule something.

**Comment by tobizk04** [comment]
Hi Jerome (@jeromelaban),
Six months ago, we discussed support for Wasm multithreading with .NET 10 and the Uno platform 6.2 or higher.
What is the current plan?

When will the Uno platform support multithreading for Wasm again?
And will it also be experimental or production-ready?

.NET 8 will reach the end of its lifecycle next year, and we need an alternative.
